### PR TITLE
Support compile time warnings for the `current::strings::Printf`

### PR DESCRIPTION
**Pull request by grixa**
Use `__attribute__(__format__,...)` to enable printf-like warnings.
Unfortunately, there is no an alternative for Windows: msvc2015 has `SAL Annotations` and `_Printf_format_string_` option, but it works only with `/analyze` option (or through `Run Code Analysis on Solution`).

### CODE REVIEW

File: Bricks/net/http/test.cc [hunk 1]
```diff
@@ -441,10 +441,10 @@ TEST(PosixHTTPServerTest, ChunkedSmoke) {
     const auto offset = body.length() - length;
     for (size_t chunks = length; chunks; chunks = chunks * 2 / 3) {
       std::string chunked_body;
-      for (size_t i = 0; i < chunks; ++i) {
-        const size_t start = offset + length * i / chunks;
-        const size_t end = offset + length * (i + 1) / chunks;
-        chunked_body += current::strings::Printf("%X\r\n", end - start) + body.substr(start, end - start);
+      for (uint64_t i = 0; i < chunks; ++i) {
```
Comment (dkorolev) [inline Bricks/net/http/test.cc]: But `chunks` is `size_t`, why this change?

File: Bricks/net/http/test.cc [hunk 2]
```diff
@@ -441,10 +441,10 @@ TEST(PosixHTTPServerTest, ChunkedSmoke) {
...
+        const size_t start = offset + static_cast<size_t>(length * i / chunks);
+        const size_t end = offset + static_cast<size_t>(length * (i + 1) / chunks);
+        chunked_body += current::strings::Printf("%lX\r\n", end - start) + body.substr(start, end - start);
```
Comment (dkorolev) [inline Bricks/net/http/test.cc]: Perhaps just cast to a `uint64_t` here, as the 2nd parameter to `Printf`?

File: Bricks/strings/printf.h [hunk 3]
```diff
@@ -33,6 +33,9 @@ SOFTWARE.
 namespace current {
 namespace strings {
 
+#ifndef CURRENT_WINDOWS
```
Comment (dkorolev) [inline Bricks/strings/printf.h]: I'm cool with this staying for `CURRENT_WINDOWS`. @mzhurovich, you too? :-)

Comment (dkorolev) [inline Bricks/strings/printf.h]: Oh, I'm an idiot. This is *not* for Windows.

Perhaps guard it by compiler type/version then? An `#ifdef` for whether we're under `g++` or `clang++`?

CC @mzhurovich

Comment (mzhurovich) [inline Bricks/strings/printf.h]: I suppose it should be supported by any version of GCC/Clang that knows about c++11, so adding some specific version guards may be an overkill.

File: Bricks/net/http/test.cc (hunk 1 above)
Comment (mzhurovich) [inline Bricks/net/http/test.cc]: As the commit message says - "integer overflow" :)

File: Bricks/strings/printf.h (hunk 3 above)
Comment (dkorolev) [inline Bricks/strings/printf.h]: Hate to sound like a prick, but there are other compilers. Perhaps introduce a top-level `CURRENT_GCC_OR_CLANG` into `port.h` and guard by it here?

File: Bricks/net/http/test.cc (hunk 1 above)
Comment (dkorolev) [inline Bricks/net/http/test.cc]: OH SHI~!

Then perhaps make the outer look `uint64_t` too?

Sorry, still hungry and still waking up after a lava hike yesterday.

Comment (dkorolev) [inline Bricks/net/http/test.cc]: *the outer loop

File: Bricks/strings/printf.h (hunk 3 above)
Comment (mzhurovich) [inline Bricks/strings/printf.h]: Ok, `__attributes__` is a GNU extension and should be supported by a GNU compiler, so the right ifdef here is
```
#ifdef __GNUC__
```

Comment (grixa) [inline Bricks/strings/printf.h]: Thanks! Fixed in the a33e245

File: Bricks/net/http/test.cc (hunk 1 above)
Comment (grixa) [inline Bricks/net/http/test.cc]: You are right, it's better to use `uint64_t` everywhere around to make the code look less confusing. 59ea5d0

**Comment by dkorolev** [comment]
Thank you Grisha and Max! Merging.
